### DISCUSSION TITLE
seccomp and apparmor

### DISCUSSION THREAD

**Discussion by ishfx** [photoprism/photoprism#1273] [Ideas]
Hello,

I didn't find any answer in the documentation to my question : 
Why does photoprism need **seccomp** and **apparmor** to be **unconfined** ?

Does is use ALL the syscalls (likely not) ?
Does it need complete rwx access to all files/network ?

Thanks in advance.

**Comment by lastzero** [comment]
No, it doesn't need all. Some users reported issues though and were not able to configure their kernel security properly. You're welcome to contribute docs, test different configurations, and help with answering support requests so that all users can enjoy the benefits of strong kernel security for their personal home server.

**Comment by ishfx** [comment]
Alright, thanks ! I'll run some tests with docker's default security profile and see which part of photoprism breaks.
In the meantime, i think the issue can be closed.

**Comment by alexisLefebvre** [comment]
I run PhotoPrism without `security_opt` on Debian 10 and I don't have any issue (I installed it before that option was added). Please reopen this issue so that we can keep track of it.

Maybe these lines should be commented with an explanation: https://github.com/photoprism/photoprism/blob/2fb7a4d7ea151d86ca6bb2af48c8c03a8c3bfa0e/docker/examples/docker-compose.yml#L40-L42

**Comment by lastzero** [comment]
Should we move this to our docs repo? It's really not a feature of PhotoPrism nor do I think the average home user needs enterprise grade Linux Kernel security. Most users don't even know how file system permissions work. It's an advanced topic for those who want / need it.

**Reply by mreilaender** [reply]
I couldn't disagree more. Providing an example of running your app in docker with some default security features disabled WITHOUT giving an explanation why, gives the wrong impression to your users. Yes most of your users might not care but this should NOT be used as an excuse to rectify "bad" application design.

Also you didn't really answer the question of "why does photoprism need those security features to be disabled". I get that you might not be able to reproduce those issues but IMO "some users had problems" should not be used as a rectification to put disabling security features in an official documentation.

**Reply by lastzero** [reply]
You will generally not need this. However, we cannot offer personal installation assistance to every user and for every operating system that has problems. For home users, I don't think these settings significantly increase the risk. If you have other information that you can underpin with numbers, then go ahead.

**Reply by lastzero** [reply]
To prevent other users from misinterpreting your comment, I would like to add that the software directly maintained by us - i.e. PhotoPrism itself - has never had any issues with kernel security settings. The cases I know of involved **MariaDB** (got fixed with a MariaDB image and/or a Docker update, from what I remember) as well as **Darktable**, which itself depends on a long list of other libraries, including some for hardware acceleration that are likely pretty low-level and may use cutting edge functionality not foreseen by the developers of some Linux distributions. Given that, I don't think it's fair (or worth anyone's time) to blame us so aggressively, right?

**Reply by mreilaender** [reply]
I'm sorry if my comment came out as mean / aggressive, that wasn't my intention.

Nevertheless my point was that putting disabling of security features into the first usage example of the app into the official documentation doesn't leave a very good impression. I would suggest to remove the `seccomp:unconfined` and `apparmor:unconfined ` from the official docker usage example and move it into a troubleshooting section. And also put a note there informing the user that this renders the container more vulnerable in terms of security.

**Reply by lastzero** [reply]
Thank you! Regarding your feedback: My personal impression is that anyone who understands security and cares about a secure system (having an updated / properly secured Linux) will remove those lines when customizing the Docker Compose file (which is the first thing you should do according to our documentation).

As maintainers of free software, time really is our scarcest resource. So anything that takes away that time negatively impacts our work-life balance, security, and new feature development, e.g. home users panicking after reading that this makes them more vulnerable (even though they're still using kernel version 3.10 and/or haven't installed any security updates), or not reading the troubleshooting section and then contacting our "customer support" because something isn't working.

We know there are many things we should be doing in addition to what we're already doing, but adding even more work to our plate doesn't help the situation and may even have a negative impact on security because it takes away time we would have otherwise invested in updating our libraries and doing proper testing.

That being said, I would appreciate some help with improving and expanding the related documentation (as well as testing and improving our software in general):

- https://docs.photoprism.app/getting-started/troubleshooting/docker/#kernel-security

For example, it would be great to have an overview of the Docker and Linux versions that may require custom settings (as a first step, you could check open and closed GitHub issues for related reports), so that users can easily see if the most secure settings work for them without doing time-consuming tests or asking us for personal advice! :heart:
